File: ScaleformUI_Csharp/Menus/Pause Menus/PauseMenu/Items/TabLeftItem.cs
using ScaleformUI.Elements;
using ScaleformUI.Menu;
using ScaleformUI.Menus;

namespace ScaleformUI.PauseMenu
{
    public enum LeftItemType
    {
        Empty,
        Info,
        Statistics,
        Settings,
        Keymap
    }

    public enum LeftItemBGType
    {
        Full,
        Masked,
        Resized
    }

    public delegate void IndexChangeEvent(SettingsItem item, int index);
    public delegate void ActivatedEvent(TabLeftItem item, int index);

    public class TabLeftItem : PauseMenuItem
    {
        internal UIMenuItem _internalItem { get; set; }
        private bool enabled = true;
        private SColor mainColor = SColor.HUD_Pause_bg;
        private SColor highlightColor = SColor.HUD_White;
        private string textTitle;
        private string _label = "";
        internal string _formatLeftLabel = "";
        private string keymapRightLabel_1;
        private string keymapRightLabel_2;
        public string TextureDict { get; private set; }
        public string TextureName { get; private set; }
        public LeftItemBGType LeftItemBGType { get; private set; }

        internal ItemFont _labelFont = ScaleformFonts.CHALET_LONDON_NINETEENSIXTY;
        internal ItemFont _rightLabelFont = ScaleformFonts.CHALET_LONDON_NINETEENSIXTY;
        public LeftItemType ItemType { get; internal set; }
        public new string Label
        {
            get => _label;
            set
            {
                _label = value;
                _formatLeftLabel = value.StartsWith("~") ? value : "~s~" + value;
                _formatLeftLabel = !Enabled ? _formatLeftLabel.ReplaceRstarColorsWith("~c~") : Selected ? _formatLeftLabel.Replace("~w~", "~l~").Replace("~s~", "~l~") : _formatLeftLabel.Replace("~l~", "~s~");
            }
        }
        public SColor MainColor { get => mainColor; set => mainColor = value; }
        public SColor HighlightColor { get => highlightColor; set => highlightColor = value; }
        public bool Enabled
        {
            get => enabled;
            set
            {
                enabled = value;
                if (!value)
                    _formatLeftLabel = _formatLeftLabel.ReplaceRstarColorsWith("~c~");
                else
                    Label = _label;

                if (ParentTab != null && ParentTab.Visible)
                {
                    int it = ParentTab.LeftColumn.Items.IndexOf(this);
                    ParentTab.UpdateSlot(PM_COLUMNS.LEFT, it);
                }
            }
        }

        public bool Hovered { get; internal set; }

        public override bool Selected
        {
            get => selected;
            set
            {
                selected = value;
                if (value)
                    _formatLeftLabel = _formatLeftLabel.Replace("~w~", "~l~").Replace("~s~", "~l~");
                else
                    _formatLeftLabel = _formatLeftLabel.Replace("~l~", "~s~");
                int it = ParentTab.LeftColumn.Items.IndexOf(this);
                ParentTab.UpdateSlot(PM_COLUMNS.LEFT, it);
            }
        }

        public int ItemIndex { get; internal set; }
        public List<PauseMenuItem> ItemList { get; set; } = new List<PauseMenuItem>();
        private bool selected;

        public string RightTitle
        {
            get => textTitle;
            set
            {
                textTitle = value;
                if (ParentTab != null && ParentTab.Visible)
                {
                    int it = ParentTab.LeftColumn.Items.IndexOf(this);
                    ParentTab.UpdateSlot(PM_COLUMNS.LEFT, it);
                }
            }
        }

        public string KeymapRightLabel_1
        {
            get => keymapRightLabel_1;
            set
            {
                keymapRightLabel_1 = value;
                if (ParentTab != null && ParentTab.Visible)
                {
                    int it = ParentTab.LeftColumn.Items.IndexOf(this);
                    ParentTab.UpdateSlot(PM_COLUMNS.LEFT, it);
                }
            }
        }
        public string KeymapRightLabel_2
        {
            get => keymapRightLabel_2;
            set
            {
                keymapRightLabel_2 = value;
                if (ParentTab != null && ParentTab.Visible)
                {
                    int it = ParentTab.LeftColumn.Items.IndexOf(this);
                    ParentTab.UpdateSlot(PM_COLUMNS.LEFT, it);
                }
            }
        }

        public event ActivatedEvent OnActivated;
        public new SubmenuTab ParentTab { get; set; }

        public TabLeftItem(string label, LeftItemType type) : this(label, type, SColor.HUD_Pause_bg, SColor.HUD_White) { }
        public TabLeftItem(string label, LeftItemType type, SColor mainColor, SColor highlightColor) : base(label)
        {
            Label = label;
            ItemType = type;
            MainColor = mainColor;
            HighlightColor = highlightColor;
            _internalItem = new UIMenuItem(label, "", SColor.HUD_Pause_bg, SColor.HUD_White);
        }
        public TabLeftItem(string label, LeftItemType type, ItemFont labelFont, SColor mainColor, SColor highlightColor) : base(label)
        {
            Label = label;
            ItemType = type;
            MainColor = mainColor;
            HighlightColor = highlightColor;
            _labelFont = labelFont;
            _internalItem = new UIMenuItem(label, "", SColor.HUD_Pause_bg, SColor.HUD_White);
        }

        /// <summary>
        /// Image suggested not bigger than 720 x 576
        /// </summary>
        /// <param name="txd"></param>
        /// <param name="txn"></param>
        public void UpdateBackground(string txd, string txn, LeftItemBGType resizeType)
        {
            TextureDict = txd;
            TextureName = txn;
            LeftItemBGType = resizeType;
            if (ParentTab != null && ParentTab.Visible && ParentTab.Parent != null && ParentTab.Parent.Visible && this.ItemType != LeftItemType.Empty)
            {
                int it = ParentTab.LeftColumn.Items.IndexOf(this);
                ParentTab.UpdateSlot(PM_COLUMNS.LEFT, it);
            }
        }

        public void AddItem(PauseMenuItem item)
        {
            item.ParentLeftItem = this;
            item.ParentTab = ParentTab;
            ItemList.Add(item);
        }

        internal void Activated()
        {
            OnActivated?.Invoke(this, ParentTab.LeftColumn.Items.IndexOf(this));
        }
    }
}

File: ScaleformUI_Lua/src/Menus/PauseMenus/PauseMenu/Items/TabLeftItem.lua
TabLeftItem = {}
TabLeftItem.__index = TabLeftItem
setmetatable(TabLeftItem, { __index = PauseMenuItem })
TabLeftItem.__call = function()
    return "TabLeftItem"
end

---@class TabLeftItem
---@field public ItemType number
---@field public Focused boolean
---@field public MainColor SColor
---@field public HighlightColor SColor
---@field public Highlighted boolean
---@field public ItemIndex number
---@field public ItemList table<SettingsItem|SettingsListItem|SettingsProgressItem|SettingsCheckboxItem|SettingsSliderItem|KeymapItem>
---@field public RightTitle string
---@field private _enabled boolean
---@field private _hovered boolean
---@field private _selected boolean
---@field public KeymapRightLabel_1 string
---@field public KeymapRightLabel_2 string
---@field public Index number
---@field public AddItem fun(self:TabLeftItem, item:SettingsItem|SettingsListItem|SettingsProgressItem|SettingsCheckboxItem|SettingsSliderItem|KeymapItem|PauseMenuItem)
---@field public Enabled fun(self:TabLeftItem, enabled:boolean):boolean
---@field public Hovered fun(self:TabLeftItem, hover:boolean):boolean
---@field public Selected fun(self:TabLeftItem, selected:boolean):boolean
---@field public OnIndexChanged fun(item:TabLeftItem, index:number)
---@field public OnActivated fun(item:TabLeftItem, index:number)

function TabLeftItem.New(label, _type, mainColor, highlightColor, labelFont)
    local base = PauseMenuItem.New(label, labelFont)
    base.ItemType = _type
    base.Focused = false
    base.MainColor = mainColor or SColor.HUD_Pause_bg
    base.HighlightColor = highlightColor or SColor.HUD_White
    base.Highlighted = false
    base.ItemList = {} --[[@type table<SettingsItem|SettingsListItem|SettingsProgressItem|SettingsCheckboxItem|SettingsSliderItem|KeymapItem|PauseMenuItem>]]
    base._enabled = true
    base._hovered = false
    base._formatLeftLabel = __formatLeftLabel or ""
    base.KeymapRightLabel_1 = ""
    base.KeymapRightLabel_2 = ""
    base.TextureDict = ""
    base.TextureName = ""
    base.TextTitle = ""
    base.LeftItemBGType = 0
    base.OnActivated = function(item, index)
    end
    return setmetatable(base, TabLeftItem)
end

---Add item to the tab list
---@param item SettingsItem|SettingsListItem|SettingsProgressItem|SettingsCheckboxItem|SettingsSliderItem|KeymapItem|PauseMenuItem
function TabLeftItem:AddItem(item)
    item.ParentLeftItem = self
    item.ParentTab = self.ParentTab
    self.ItemList[#self.ItemList + 1] = item
end

function TabLeftItem:Label(label)
    if label ~= nil then
        self.label = label
        if self.ParentTab ~= nil and self.ParentTab.Visible then
            local leftItem = IndexOf(self.ParentTab.LeftColumn.Items, self)
            self.ParentTab.LeftColumn:UpdateSlot(leftItem)
        end
    else
        return self.label
    end
end

function TabLeftItem:Enabled(enabled)
    if enabled ~= nil then
        self._enabled = enabled
        if self.ParentTab ~= nil and self.ParentTab.Visible then
            local leftItem = IndexOf(self.ParentTab.LeftColumn.Items, self)
            self.ParentTab.LeftColumn:UpdateSlot(leftItem)
        end
    else
        return self._enabled
    end
end

function TabLeftItem:Hovered(hover)
    if hover ~= nil then
        self._hovered = hover
    else
        return self._hovered
    end
end

function TabLeftItem:Selected(selected)
    if selected ~= nil then
        self._selected = selected
        if self.ParentTab ~= nil and self.ParentTab.Visible then
            local leftItem = IndexOf(self.ParentTab.LeftColumn.Items, self)
            self.ParentTab.LeftColumn:UpdateSlot(leftItem)
        end
    else
        return self._selected
    end
end

function TabLeftItem:RightTitle(rtit)
    if rtit ~= nil then
        self.TextTitle = rtit
        if self.ParentTab ~= nil and self.ParentTab.Visible then
            local leftItem = IndexOf(self.ParentTab.LeftColumn.Items, self)
            self.ParentTab.LeftColumn:UpdateSlot(leftItem)
        end
    else
        return self.TextTitle
    end
end

function TabLeftItem:KeymapRightLabel_1(rtit)
    if rtit ~= nil then
        self.KeymapRightLabel_1 = rtit
        if self.ParentTab ~= nil and self.ParentTab.Visible then
            local leftItem = IndexOf(self.ParentTab.LeftColumn.Items, self)
            self.ParentTab.LeftColumn:UpdateSlot(leftItem)
        end
    else
        return self.KeymapRightLabel_1
    end
end

function TabLeftItem:KeymapRightLabel_2(rtit)
    if rtit ~= nil then
        self.KeymapRightLabel_2 = rtit
        if self.ParentTab ~= nil and self.ParentTab.Visible then
            local leftItem = IndexOf(self.ParentTab.LeftColumn.Items, self)
            self.ParentTab.LeftColumn:UpdateSlot(leftItem)
        end
    else
        return self.KeymapRightLabel_2
    end
end

-- legacy and not supported anymore
-- function TabLeftItem:UpdateBackground(txd, txn, resizeType)
--     self.TextureDict = txd
--     self.TextureName = txn
--     self.LeftItemBGType = resizeType
--     if self.ParentTab ~= nil and self.ParentTab.Visible then
--         local leftItem = IndexOf(self.ParentTab.LeftColumn.Items, self)
--         self.ParentTab.LeftColumn:UpdateSlot(leftItem)
--     end
-- end
